ostd/src/device/dispatcher/io_mem.rs

```diff
@@ -0,0 +1,234 @@
+            if allocator.alloc_specific(range).is_err() {
+                return None;
+            }
+
+            // SAFETY: The created IoMem is guaranteed not to access physical memory or system device I/O
+            unsafe {
+                result = Some(IoMem::new(
+                    start_addr..end_addr,
+                    PageFlags::RW,
+                    CachePolicy::Uncacheable,
+                ))
+            }
+            break;
+        }
+        result
+    }
+
+    /// Remove access to a specific memory IO range.  
+    ///
+    /// All drivers in OSTD must use this method to prevent peripheral drivers from accessing illegal memory IO range.
+    pub(crate) fn remove(&self, range: Range<usize>) {
+        for allocator in self.allocators.get().unwrap().iter() {
+            let allocator_range = allocator.fullrange();
+            if allocator_range.start >= range.end || allocator_range.end <= range.start {
+                continue;
+            }
+
+            let start_addr = range.start;
+            let end_addr = range.end;
+            debug!("Removing MMIO range:{:x}..{:x}", start_addr, end_addr);
+
+            // Remove I/O memory.
+            let _ = allocator.alloc_specific(range);
+            return;
+        }
+    }
+
+    /// Recycle an MMIO range.
+    #[allow(dead_code)]
+    pub(crate) fn recycle(&self, range: Range<usize>) {
+        for allocator in self.allocators.get().unwrap().iter() {
+            let allocator_range = allocator.fullrange();
+            if allocator_range.start >= range.end || allocator_range.end <= range.start {
+                continue;
+            }
+
+            let start_addr = range.start;
+            let end_addr = range.end;
+            debug!("Recycling MMIO range:{:x}..{:x}", start_addr, end_addr);
+
+            // Recycle I/O memory.
+            allocator.free(range);
+            return;
+        }
+    }
+
+    /// Initialize usable memory I/O region.
+    ///
+    /// # Safety
+    ///
+    /// User must ensure the range doesn't belong to physical memory.
+    unsafe fn init(&self, ranges: Vec<Range<usize>>) {
+        self.allocators.call_once(|| {
+            info!(
+                "Initialize the IO memory dispatcher with range:{:x?}",
+                ranges
+            );
+            let mut allocators = Vec::with_capacity(ranges.len());
+            for range in ranges {
+                allocators.push(VirtAddrAllocator::new(range));
+            }
+            allocators
+        });
+    }
+
+    const fn new() -> Self {
+        Self {
+            allocators: Once::new(),
+        }
+    }
+}
+
+/// The I/O Memory dispatcher of the system.
+pub static IO_MEM_DISPATCHER: IoMemDispatcher = IoMemDispatcher::new();
+
+pub(crate) fn init() {
+    // Initialize the allocator of the IoMemDispatcher based on the system memory info.
+    #[cfg(target_arch = "x86_64")]
+    init_dispatcher(crate::boot::memory_regions().to_owned());
```

Comment:
It makes no sense to query the memory regions here. Why not query them in `init_dispatcher`?